File: server/infra/telemetry/metrics.py
"""Predefined OTel metric instruments for /bridge/chat.

Instruments are created at module import time using the current MeterProvider
(initially NoOp). After ``init_telemetry()`` installs the real provider,
``ensure_instruments()`` re-creates them so they record to Redis.

When OTLP is disabled the provider stays NoOp, so ``.add()`` / ``.record()``
calls are zero-overhead no-ops.
"""

from __future__ import annotations

from opentelemetry.metrics import (
    Counter,
    Histogram,
    UpDownCounter,
    Meter,
    get_meter,
)

# Histogram bucket boundaries
REQUEST_DURATION_BUCKETS = (
    0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1.0, 2.5, 5.0, 10.0,
)
STREAM_DURATION_BUCKETS = (1.0, 5.0, 10.0, 30.0, 60.0, 120.0, 300.0)


def _create_instruments(meter: Meter) -> tuple[Counter, Histogram, Histogram, UpDownCounter]:
    return (
        meter.create_counter(
            name="bridge.chat.requests",
            description="/bridge/chat 请求总数",
        ),
        meter.create_histogram(
            name="bridge.chat.request.duration",
            description="/bridge/chat 首字节耗时",
            unit="s",
        ),
        meter.create_histogram(
            name="bridge.chat.stream.duration",
            description="SSE 流持续时长",
            unit="s",
        ),
        meter.create_up_down_counter(
            name="bridge.chat.active_streams",
            description="当前活跃 SSE 流数量",
        ),
    )


# Initialise with NoOp meter (safe at import time)
_meter = get_meter("astron_claw")
chat_request_total, chat_request_duration, chat_stream_duration, chat_active_streams = (
    _create_instruments(_meter)
)


def ensure_instruments() -> None:
    """Re-create instruments using the current (real) MeterProvider.

    Must be called after ``init_telemetry()`` so the real MeterProvider is active.
    Idempotent — safe to call multiple times.
    """
    global chat_request_total, chat_request_duration
    global chat_stream_duration, chat_active_streams

    meter: Meter = get_meter("astron_claw")
    chat_request_total, chat_request_duration, chat_stream_duration, chat_active_streams = (
        _create_instruments(meter)
    )


def _token_prefix(token: str) -> str:
    """Return first 10 chars + '...' for token label."""
    return f"{token[:10]}..." if len(token) > 10 else token

File: backend/internal/infra/telemetry/metrics.go
package telemetry

import (
	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/metric"
)

// Histogram bucket boundaries
var (
	RequestDurationBuckets = []float64{0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1.0, 2.5, 5.0, 10.0}
	StreamDurationBuckets  = []float64{1.0, 5.0, 10.0, 30.0, 60.0, 120.0, 300.0}
)

// Global instruments — initialized with NoOp meter, re-created after init_telemetry.
var (
	ChatRequestTotal    metric.Int64Counter
	ChatRequestDuration metric.Float64Histogram
	ChatStreamDuration  metric.Float64Histogram
	ChatActiveStreams   metric.Int64UpDownCounter
)

func init() {
	EnsureInstruments()
}

// EnsureInstruments creates or re-creates OTel instruments using the current MeterProvider.
func EnsureInstruments() {
	meter := otel.Meter("astron_claw")

	var err error
	ChatRequestTotal, err = meter.Int64Counter(
		"bridge.chat.requests",
		metric.WithDescription("/bridge/chat request total"),
	)
	if err != nil {
		ChatRequestTotal, _ = meter.Int64Counter("bridge.chat.requests")
	}

	ChatRequestDuration, err = meter.Float64Histogram(
		"bridge.chat.request.duration",
		metric.WithDescription("/bridge/chat first-byte latency"),
		metric.WithUnit("s"),
	)
	if err != nil {
		ChatRequestDuration, _ = meter.Float64Histogram("bridge.chat.request.duration")
	}

	ChatStreamDuration, err = meter.Float64Histogram(
		"bridge.chat.stream.duration",
		metric.WithDescription("SSE stream duration"),
		metric.WithUnit("s"),
	)
	if err != nil {
		ChatStreamDuration, _ = meter.Float64Histogram("bridge.chat.stream.duration")
	}

	ChatActiveStreams, err = meter.Int64UpDownCounter(
		"bridge.chat.active_streams",
		metric.WithDescription("Current active SSE stream count"),
	)
	if err != nil {
		ChatActiveStreams, _ = meter.Int64UpDownCounter("bridge.chat.active_streams")
	}
}

// TokenPrefix returns first 10 chars + "..." for token labels.
func TokenPrefix(token string) string {
	if len(token) > 10 {
		return token[:10] + "..."
	}
	return token
}
